Frameworks/UIKit/_UIGestureCoordinator.mm

```diff
@@ -0,0 +1,457 @@
+#import "_UIGestureCoordinator.h"
+
+#import <UIKit/UIGestureRecognizer.h>
+#import <UIKit/UIView.h>
+#import <UIKit/UITouch.h>
+#import <UIKit/UIGestureRecognizerDelegate.h>
+#import <UIKit/UIGestureRecognizerSubclass.h>
+
+#import "UIViewInternal.h"
+#import "UIGestureRecognizerInternal.h"
+#import "_UIDirectManipulationRecognizer.h"
+
+static const wchar_t* TAG = L"_UIGestureCoordinator";
+
+static const bool DEBUG_ALL = false;
+static const bool DEBUG_GESTURES = DEBUG_ALL || true;
+
+static _UIGestureCoordinator* g_UIGestureCoordinator = nil;
+
+@implementation _UIGestureCoordinator {
+    // internal flag indicating all gestures has been
+    // reset, so when processing new touch, we need
+    // rebuild the gesture list
+    BOOL _resetAllTrackingGestures;
+
+    // regular gesture recognizer list
+    // Prefer NSMutableSortedSet, but we don't support yet,
+    // using NSMutableArray instead, same reason for _dManipRecognizers
+    // when _resetAllTrackingGestures is set to YES, during touch processing
+    // we will rebuild regular gesture list
+    NSMutableArray* _regularRecognizers;
+
+    // system DManip gesture recognizer list
+    // when _resetAllTrackingGestures is set to YES, during touch processing
+    // we will rebuild DManip gesture list
+    NSMutableArray* _dManipRecognizers;
+
+    // gesture list used for booking up gestures to be removed in one Touch
+    // processing cycle. When processing a Touch event, some gestures can be
+    // cancelled or be recognized, thsu become ended. These gestures will be
+    // stored in this variable and then clean up. So when processing next touch
+    // we don't have to enumerate through ended gesture, improve enumeration efficience
+    NSMutableSet* _gesturesToRemove;
+
+    // gesture list used for booking gestures in possible state.  when one candidate
+    // gesture in _regularRecognizers becomes active we need enumerate this list to
+    // make sure that if all gestures in this list can be fired concurrently, if not,
+    // they will be transit to cancelled state and removed from this list.
+    // it improves gesture processing efficiency
+    NSMutableSet* _gesturesInPossibleState;
+}
+
++ (instancetype)singleton {
+    if (!g_UIGestureCoordinator) {
+        g_UIGestureCoordinator = [_UIGestureCoordinator new];
+    }
+
+    return g_UIGestureCoordinator;
+}
+
+- (instancetype)init {
+    if (self = [super init]) {
+        _resetAllTrackingGestures = YES;
+        _gesturesToRemove = [NSMutableSet new];
+        _gesturesInPossibleState = [NSMutableSet new];
+        _regularRecognizers = [NSMutableArray new];
+        _dManipRecognizers = [NSMutableArray new];
+    }
+
+    return self;
+}
+
+// test override to provide custom gestures
+- (void)buildTrackingGesturesListWith:(NSMutableArray*)regularRecognizers
+                     PossibleGestures:(NSMutableSet*)gesturesInPossibleState
+                        DManipGesture:(NSMutableArray*)dManipRecognizers {
```

Comment:
it did. 

---
In reply to: [100622933](https://github.com/Microsoft/WinObjC/pull/1909#discussion_r100622933) [](ancestors = 100622933)